tokio/src/stream/throttle.rs

```diff
@@ -46,14 +46,29 @@ pin_project! {
     }
 }
 
-// XXX: are these safe if `T: !Unpin`?
-impl<T: Unpin> Throttle<T> {
+impl<T> Throttle<T> {
+    fn get_sleep_mut(self: Pin<&mut Self>) -> Option<Pin<&mut Sleep>> {
+        unsafe {
+            // SAFETY: We project here to the inner Sleep object. Because we
+            // allow access only via a Pin<&mut Sleep>, our returned reference
+            // does not allow the pinning guarantees to be violated.
```

Comment:
I don't think this is correct, pin-project(-lite) assume non-`#[pin]` field is not structurally pinned fields (i.e., never pinned) and implement `Unpin` based on that assumption.
https://docs.rs/pin-project/1.0.1/pin_project/attr.pin_project.html#safety

> To enforce this, this attribute will automatically generate an `Unpin` implementation for you, which will require that all structurally pinned fields be `Unpin`.

In this case, `Unpin` implementation like the following is generated:
```rust
impl<T> Unpin for Throttle<T> where T: Unpin {}
```
This violates the following safety requirements if `Sleep` is not `Unpin`. (`delay` field is structurally pinned fields.)
https://doc.rust-lang.org/nightly/core/pin/index.html#pinning-is-structural-for-field

> 1. The struct must only be `Unpin` if all the structural fields are `Unpin`. 

So I think the correct solution here is to add a `#[pin]` attribute to `delay` field and use `.as_pin_mut()` to convert `Pin<&mut Option<Sleep>>` to `Option<Pin<&mut Sleep>>`.

Reply:
I agree. This unsafe block makes the `Unpin` impl block generated by `pin-project-lite` incorrect.

Reply:
Interesting! Is there a way to add a must-compile-fail test to help catch these things, I wonder?

Reply:
If you extend `assert_value!` to check for `Unpin` too, you can add an `!Unpin` assertion for the type to the [`assert_send_sync.rs`][1] test.
```rust
assert_value!(tokio::stream::Throttle<i32>: !Unpin);
```
This uses an `Unpin` generic type to verify that it is `!Unpin` even if the argument is `Unpin`.

[1]: https://github.com/tokio-rs/tokio/blob/master/tokio/tests/async_send_sync.rs

Reply:
Hmm, there's no `!Unpin` defined for the `assert_value!` macro currently: https://github.com/tokio-rs/tokio/blob/02311dcfa13d719c3f591c922f8a719d7c954ef0/tokio/tests/async_send_sync.rs#L60

That being said `AmbiguousIfUnpin` is already defined so it looks straightforward enough to add it.

Reply:
Backing out `!Unpin` for now (boxing in Sleep) so this can go out in 0.3, so this is now moot.